packages/google_maps_flutter/google_maps_flutter_web/lib/src/overlays.dart

```diff
@@ -23,72 +23,78 @@ class TileOverlaysController {
     _visibleTileOverlays.insert(index, tileOverlayController);
   }
 
-  void _removeZSorted(TileOverlayController tileOverlayController) {
+  // Removes `tileOverlayController` from the list of visible overlays.
+  void _remove(TileOverlayController tileOverlayController) {
     final int index = _visibleTileOverlays.indexOf(tileOverlayController);
+    if (index < 0) {
+      return;
+    }
+
     googleMap.overlayMapTypes!.removeAt(index);
     _visibleTileOverlays.removeAt(index);
   }
 
   /// Adds new [TileOverlay]s to this controller.
   ///
   /// Wraps the [TileOverlay]s in corresponding [TileOverlayController]s.
-  void addTileOverlays(Set<TileOverlay> tileOverlays) {
-    for (final TileOverlay tileOverlay in tileOverlays) {
-      final TileOverlayController controller = TileOverlayController()
-        ..update(tileOverlay);
-      _tileOverlays[tileOverlay.tileOverlayId] = controller;
-
-      if (tileOverlay.visible) {
-        _insertZSorted(controller);
-      }
+  void addTileOverlays(Set<TileOverlay> tileOverlaysToAdd) {
+    tileOverlaysToAdd.forEach(_addTileOverlay);
+  }
+
+  void _addTileOverlay(TileOverlay tileOverlay) {
+    final TileOverlayController controller = TileOverlayController(
+      tileOverlay: tileOverlay,
+    );
+    _tileOverlays[tileOverlay.tileOverlayId] = controller;
+
+    if (tileOverlay.visible) {
+      _insertZSorted(controller);
     }
   }
 
   /// Updates [TileOverlay]s with new options.
   void changeTileOverlays(Set<TileOverlay> tileOverlays) {
-    for (final TileOverlay tileOverlay in tileOverlays) {
-      final TileOverlayController controller =
-          _tileOverlays[tileOverlay.tileOverlayId]!;
-
-      final bool wasVisible = controller.tileOverlay.visible;
-      final bool zIndexChanged =
-          tileOverlay.zIndex != controller.tileOverlay.zIndex;
-      if (wasVisible && !tileOverlay.visible || zIndexChanged) {
-        _removeZSorted(controller);
-      }
-
-      controller.update(tileOverlay);
-
-      if (tileOverlay.visible) {
-        if (!wasVisible || zIndexChanged) {
-          _insertZSorted(controller);
-        } else {
-          final int i = _visibleTileOverlays.indexOf(controller);
-          googleMap.overlayMapTypes!.setAt(i, controller.gmMapType);
-        }
-      }
+    tileOverlays.forEach(_changeTileOverlay);
+  }
+
+  void _changeTileOverlay(TileOverlay tileOverlay) {
+    final TileOverlayController controller =
+        _tileOverlays[tileOverlay.tileOverlayId]!;
+
+    final bool wasVisible = controller.tileOverlay.visible;
+    final bool isVisible = tileOverlay.visible;
+    final bool moved = tileOverlay.zIndex != controller.tileOverlay.zIndex;
+
+    controller.update(tileOverlay);
+
+    if (wasVisible || moved) {
```

Comment:
This little bit of code concerns me a little bit as well, but we'll see in the real world :) I simplified the original if-else a little bit and ended up with this, and I might have over-simplified. I'll tag you whenever the first bug comes about this so you can tell me: "I TOLD YOU SO!" :p

Reply:
Oh actually you know what, rereading this more carefully I think it does warrant another pass. For example, there actually seems to be a bug (that I almost doubt anyone would hit) where if you change the Z index of a layer that's not visible, it becomes visible. (Purely based on my reading of the code; I haven't run a test, so I might be wrong.)

And, since this is in a loop, the impact of remove/insert vs. set for the stable z-index case, although probably still a premature optimization, could be compounded if it were ever actually an issue.

Reply:
Basically the fix if we want to do it this way is just to remove the `moved` condition entirely, but you know what I'd do if it were up to me.